extractor/docker/docker.go

```diff
@@ -145,21 +164,41 @@ func (d Extractor) createRegistryClient(ctx context.Context, domain string) (*re
 }
 
 func (d Extractor) SaveLocalImage(ctx context.Context, imageName string) (io.Reader, error) {
-	var err error
-	r := d.Cache.Get(imageName)
-	if r == nil {
-		// Save the image
-		r, err = d.saveLocalImage(ctx, imageName)
+	var storedReader io.Reader
+
+	var storedImage []byte
+	found, err := d.Cache.Get(kvtypes.GetItemInput{
+		BucketName: "imagebucket",
+		Key:        imageName,
+		Value:      &storedImage,
+	})
+
+	if found {
+		return bytes.NewReader(storedImage), nil
+	}
+
+	var savedImage []byte
+	if err != nil || !found {
+		storedReader, err = d.saveLocalImage(ctx, imageName)
 		if err != nil {
 			return nil, xerrors.Errorf("failed to save the image: %w", err)
 		}
-		r, err = d.Cache.Set(imageName, r)
+
+		savedImage, err = ioutil.ReadAll(storedReader)
```

Comment:
Some images are so large. I'd like to avoid copying an entire layer into memory. It may be better to extract only the necessary files from the image.

Reply:
Yeah I agree. Maybe we should add a TODO here and leave it for a future improvement as this PR is already getting big. What do you think?

Reply:
What do you think about this change?

```diff
--git a/extractor/docker/docker.go b/extractor/docker/docker.go
index bfdd958..10d5f46 100644
--- a/extractor/docker/docker.go
+++ b/extractor/docker/docker.go
@@ -186,14 +186,14 @@ func (d Extractor) SaveLocalImage(ctx context.Context, imageName string) (io.Rea
 		found = false
 	}
 
-	var savedImage []byte
+	var savedImage bytes.Buffer
 	if err != nil || !found {
 		storedReader, err = d.saveLocalImage(ctx, imageName)
 		if err != nil {
 			return nil, xerrors.Errorf("failed to save the image: %w", err)
 		}
 
-		savedImage, err = ioutil.ReadAll(storedReader)
+		_, err := io.Copy(&savedImage, storedReader)
 		if err != nil {
 			return nil, xerrors.Errorf("failed to read saved image: %w", err)
 		}
@@ -203,7 +203,7 @@ func (d Extractor) SaveLocalImage(ctx context.Context, imageName string) (io.Rea
 			return nil, err
 		}
 
-		dst := e.EncodeAll(savedImage, nil)
+		dst := e.EncodeAll(savedImage.Bytes(), nil)
 		if err := d.Cache.BatchSet(kvtypes.BatchSetItemInput{
 			BucketName: "imagebucket",
 			Keys:       []string{imageName},
@@ -213,7 +213,7 @@ func (d Extractor) SaveLocalImage(ctx context.Context, imageName string) (io.Rea
 		}
 	}
 
-	return bytes.NewReader(savedImage), nil
+	return bytes.NewReader(savedImage.Bytes()), nil
 }
 
 func (d Extractor) saveLocalImage(ctx context.Context, imageName string) (io.ReadCloser, error) {
```

Reply:
I'm not sure, but it looks the same because the destination of `io.Copy` is bytes.Buffer. After all, it has all bytes in memory. To be honest, I don't have a way to sort it out now 😄

Reply:
Yeah I realized that after I made this comment 😄 I was going to play with `io.Pipe` like this https://garbagecollected.org/2015/05/30/io-with-go-io-pipe/ thoughts?